### PR TITLE
Miscellaneous lazy preprocessor improvements

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

Various improvements to the preprocessor functions related to handling lazy data.
- Avoid inserting broadcasted numpy arrays into Dask graph when applying masks
- Always specify the target chunks when broadcasting
- Avoid huge chunks with derived coordinates https://github.com/SciTools/iris/issues/5457
- Ensure correct chunktype for rolling window operations https://github.com/SciTools/iris/issues/5988

Related to #674 

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by valeriupredoi** [comment]
@bouweandela this is a lovely PR - did you want to still work on it? It's been a while since it's been sitting out, let's get it in if you done working on it :beer:

**Comment by bouweandela** [comment]
It's ready for review. I would like to add more tests, but I'm not sure when I'll have time to do that.

File: esmvalcore/preprocessor/_area.py [hunk 1]
```diff
@@ -857,31 +859,45 @@ def _mask_cube(cube: Cube, masks: dict[str, np.ndarray]) -> Cube:
...
-            # (time, shape_id, depth, lat, lon)
-            if measure.ndim > 3 and result.ndim > 4:
-                data = measure.core_data()
-                data = da.expand_dims(data, axis=(1,))
-                data = da.broadcast_to(data, result.shape)
-                measure = iris.coords.CellMeasure(
+    for measure in cube.cell_measures():
+        # Cell measures that are time-dependent, with 4 dimension and
+        # an original shape of (time, depth, lat, lon), need to be
+        # broadcasted to the cube with 5 dimensions and shape
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_area.py:869]:
```suggestion
        # broadcast to the cube with 5 dimensions and shape
```

File: esmvalcore/preprocessor/_shared.py [hunk 2]
```diff
@@ -517,3 +517,85 @@ def get_dims_along_coords(
...
+    array:
+        The array to mask out.
+    dim_map : :class:`list`, :class:`tuple` etc
+        A mapping of the dimensions of *mask* to their corresponding
+        dimension in *array*. *dim_map* must be the same length as the
+        number of dimensions in *mask*. Each element of *dim_map*
+        corresponds to a dimension of *mask* and its value provides
+        the index in *array* which the dimension of *mask* corresponds
+        to, so the first element of *dim_map* gives the index of *array*
+        that corresponds to the first dimension of *mask* etc.
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_shared.py:542]: this needs to be a bit more clear:
- is it a list/tuple of ints? or of lists/tuples?
- give an example e.g. `dim_map = [2, 3, 4]` -> `array[2]` will be masked with `mask[0]`?

File: esmvalcore/preprocessor/_shared.py [hunk 3]
```diff
@@ -517,3 +517,85 @@ def get_dims_along_coords(
...
+        A copy of the input array with the mask applied.
+
+    """
+    if isinstance(array, da.Array):
+        array_chunks = array.chunks
+        # If the mask is not a Dask array yet, we make it into a Dask array
+        # before broadcasting to avoid inserting a large array into the Dask
+        # graph.
+        mask_chunks = tuple(array_chunks[i] for i in dim_map)
+        mask = da.asarray(mask, chunks=mask_chunks)
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_shared.py:556]: won't you first check if `mask` is indeed a Dask array? If it is, you'll risk rechunking with this call, as I see it

File: esmvalcore/preprocessor/_shared.py [hunk 4]
```diff
@@ -517,3 +517,85 @@ def get_dims_along_coords(
...
+    else:
+        array_chunks = None
+
+    mask = iris.util.broadcast_to_shape(
+        mask, array.shape, dim_map=dim_map, chunks=array_chunks
+    )
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_shared.py:562]: I'd be tempted to put this in the func docstring: Raises: whatever the iris utility may raise - that utility is about 90% of this function, so I'd put it in the description/API doc

File: esmvalcore/preprocessor/_shared.py [hunk 5]
```diff
@@ -517,3 +517,85 @@ def get_dims_along_coords(
...
+            if coord_dims:
+                coord = coord.copy()
+                chunks = tuple(cube_chunks[i] for i in coord_dims)
+                coord.points = coord.lazy_points().rechunk(chunks)
+                if coord.has_bounds():
+                    coord.bounds = coord.lazy_bounds().rechunk(
+                        chunks + (None,)
+                    )
+                cube.replace_coord(coord)
+    return cube
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_shared.py:601]: this looks like a costly function - is it worth its cost vs not doing anything and going with default chunks for aux coords? Also, I don't like it when we do things in our code, that should be done by Iris - this is probably something they should do at load point ie make sure the chunks are decent. Am rather not happy about this TBH, if I understand the issue correctly

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
thanks a lot @bouweandela - a few very minor points, until the last one - chunks of aux factories/coords - me not happy about that, see me comment :beer:

File: esmvalcore/preprocessor/_shared.py (hunk 3 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_shared.py:556]: It looks like this is checked on line 550 above?

File: esmvalcore/preprocessor/_shared.py (hunk 2 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_shared.py:542]: This is copy/pasted from the iris docs. I could also just link that? Anyway, this is not part of our public API.

File: esmvalcore/preprocessor/_shared.py (hunk 5 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_shared.py:601]: oh absolutely, I commend your work :medal_military: I am frowning at Iris, bud - when you plop all the other suggestions, I'll Approve, am aware this needs be here, am just not happy it's not at iris

File: esmvalcore/preprocessor/_shared.py (hunk 2 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_shared.py:542]: ha OK, pls post a link then - it's as clear as a foggy November day in the North Atlantic :grin: The module is private, but this seems to be a decent function for our public API? Up to you on that

File: esmvalcore/preprocessor/_shared.py (hunk 3 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_shared.py:556]: ah too true, me blind :man_facepalming:

File: esmvalcore/preprocessor/_shared.py (hunk 2 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_shared.py]: I changed it so the docstring now just refers to the `iris.util.broadcast_to_shape` function. That function has examples: https://scitools-iris.readthedocs.io/en/stable/generated/api/iris.util.html#iris.util.broadcast_to_shape

Comment (bouweandela) [reply esmvalcore/preprocessor/_shared.py]: > The module is private, but this seems to be a decent function for our public API?

Let's keep it private for now, I'm not sure if this is the kind of functionality ESMValCore should provide.

File: esmvalcore/preprocessor/_shared.py (hunk 4 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_shared.py:558]: brrr...I think it should, but then again, am Latin, I'd raise an exception every 10 minutes :rofl: Thanks for checking, buds!

**Review by valeriupredoi — APPROVED** [review]
very many thanks, Bouwe :beer:

**Pull request merged**